### PR TITLE
feat: Add EthGetBlockReceipts RPC method

### LINKED ISSUE #4839: Support `eth_getBlockReceipts`
# Summary

Another method is supported in Lotus, and so we should support it as well, eventually. See https://github.com/filecoin-project/lotus/issues/12429. The reference implementation is around 200 LoC.

# Completion Criteria

- [ ] Implement the method
- [ ] Add tests

# Additional Links & Resources

### PR DESCRIPTION
**Pull request by virajbhartiya**
## Summary of changes

Implemented `eth_getBlockReceipts` RPC method

## Reference issue to close (if applicable)

Closes #4839 

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation. All new code adheres to the team's [documentation standards](https://github.com/ChainSafe/forest/wiki/Documentation-practices),
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/tool/subcommands/api_cmd.rs [hunk 1]
```diff
@@ -1438,6 +1438,7 @@ fn eth_tests_with_tipset<DB: Blockstore>(store: &Arc<DB>, shared_tipset: &Tipset
             ))
             .unwrap(),
         ),
+        RpcTest::basic(EthGetBlockReceipts::request((block_hash.clone(),)).unwrap()),
```
Comment (hanabi1224) [inline src/tool/subcommands/api_cmd.rs:1441]: Any reason to use `basic` instead of `identity` here?

File: src/rpc/methods/eth.rs [hunk 2]
```diff
@@ -1259,6 +1259,48 @@ impl RpcMethod<2> for EthGetBlockByNumber {
...
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
+        (block_hash,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let ts = get_tipset_from_hash(ctx.chain_store(), &block_hash)?;
+
+        let (_, msgs_and_receipts) = execute_tipset(&ctx, &Arc::new(ts.clone())).await?;
+
+        let _state = StateTree::new_from_root(ctx.store_owned(), ts.parent_state())?;
```
Comment (hanabi1224) [inline src/rpc/methods/eth.rs:1281]: This line can be removed right?

File: src/rpc/methods/eth.rs [hunk 3]
```diff
@@ -1259,6 +1259,48 @@ impl RpcMethod<2> for EthGetBlockByNumber {
...
+
+        let mut receipts = Vec::with_capacity(msgs_and_receipts.len());
+
+        for (i, (msg, receipt)) in msgs_and_receipts.iter().enumerate() {
```
Comment (elmattic) [inline src/rpc/methods/eth.rs:1285]: You can use `into_iter()` here and avoir cloning the receipt.

**Review by elmattic — CHANGES_REQUESTED** [review]

**Comment by elmattic** [comment]
Good start!

Also, remember to update the change checklist in the PR and mark the checkboxes (e.g. update the Forest README file to reflect any visible changes for the users).

File: src/tool/subcommands/api_cmd.rs (hunk 1 above)
Comment (virajbhartiya) [reply src/tool/subcommands/api_cmd.rs:1441]: I am actually a bit confused about when to use `basic` and when to use `identity` in this case

File: src/rpc/methods/eth.rs (hunk 2 above)
Comment (virajbhartiya) [reply src/rpc/methods/eth.rs:1281]: Oh yes, I'll remove it

File: src/tool/subcommands/api_cmd.rs (hunk 1 above)
Comment (virajbhartiya) [reply src/tool/subcommands/api_cmd.rs:1441]: Just to confirm, 
`basic`: Only checks schema validity
`identity`: Checks schema validity AND exact response equality

**Comment by virajbhartiya** [comment]
I have updated the changelog and made the changes according to the reviews.

File: src/tool/subcommands/api_cmd.rs (hunk 1 above)
Comment (hanabi1224) [reply src/tool/subcommands/api_cmd.rs]: That's correct!

File: src/rpc/methods/eth.rs [hunk 4]
```diff
@@ -1259,6 +1259,46 @@ impl RpcMethod<2> for EthGetBlockByNumber {
...
+    ) -> Result<Self::Ok, ServerError> {
+        let ts = get_tipset_from_hash(ctx.chain_store(), &block_hash)?;
+
+        let (_, msgs_and_receipts) = execute_tipset(&ctx, &Arc::new(ts.clone())).await?;
+
+        let mut receipts = Vec::with_capacity(msgs_and_receipts.len());
+
+        for (i, (msg, receipt)) in msgs_and_receipts.into_iter().enumerate() {
+            let message_lookup = MessageLookup {
+                receipt: receipt.clone(),
```
Comment (elmattic) [inline src/rpc/methods/eth.rs:1285]: You're still cloning the receipt.

**Review by elmattic — CHANGES_REQUESTED** [review]

File: src/rpc/methods/eth.rs [hunk 5]
```diff
@@ -1259,6 +1259,47 @@ impl RpcMethod<2> for EthGetBlockByNumber {
...
+        (block_hash,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let ts = get_tipset_from_hash(ctx.chain_store(), &block_hash)?;
+
+        let (_, msgs_and_receipts) = execute_tipset(&ctx, &Arc::new(ts.clone())).await?;
+
+        let mut receipts = Vec::with_capacity(msgs_and_receipts.len());
+
+        for (i, (msg, receipt)) in msgs_and_receipts.into_iter().enumerate() {
+            let return_data = receipt.return_data().clone();
```
Comment (elmattic) [inline src/rpc/methods/eth.rs:1284]: You removed one clone but added another one in the process. You can avoid cloning entirely by computing `return_dec` directly.

File: src/rpc/methods/eth.rs [hunk 6]
```diff
@@ -1259,6 +1259,47 @@ impl RpcMethod<2> for EthGetBlockByNumber {
...
+            let message_lookup = MessageLookup {
+                receipt: receipt,
```
Comment (elmattic) [inline src/rpc/methods/eth.rs:1286]: You can just write `receipt` here.

**Review by elmattic — CHANGES_REQUESTED** [review]

File: src/rpc/methods/eth.rs (hunk 6 above)
Comment (elmattic) [reply src/rpc/methods/eth.rs:1286]: See: https://doc.rust-lang.org/book/ch05-01-defining-structs.html#using-the-field-init-shorthand

File: CHANGELOG.md [hunk 7]
```diff
@@ -64,6 +64,9 @@ for more details.
 - [#4706](https://github.com/ChainSafe/forest/issues/4706) Add support for the
   `Filecoin.EthSendRawTransaction` RPC method.
 
+- [#4985](https://github.com/ChainSafe/forest/pull/4985) Add support for the
+  `Filecoin.EthGetTransactionReceipt` RPC method.
```
Comment (elmattic) [inline CHANGELOG.md:68]: copy pasta error: should be `EthGetBlockReceipts`

**Review by elmattic — CHANGES_REQUESTED** [review]

File: src/rpc/methods/eth.rs (hunk 6 above)
Comment (virajbhartiya) [reply src/rpc/methods/eth.rs]: Thanks

File: src/rpc/methods/eth.rs (hunk 5 above)
Comment (virajbhartiya) [reply src/rpc/methods/eth.rs]: Oh right, I needed to do it before MessageLookup

**Comment by elmattic** [comment]
Could you resolve this lint error: https://github.com/ChainSafe/forest/actions/runs/11830899080/job/32966034648?pr=4985 and make the CI ✅ again?

Additionally, please update the README to link to the issue number instead of the PR (See https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md?plain=1#L3-L22)

Thanks!

**Comment by elmattic** [comment]
Thanks. Hey @virajbhartiya , can you investigate the CI errors? ❌ Feel free to reach me on Slack if you need some help.

**Comment by LesnyRumcajs** [comment]
@virajbhartiya It seems Forest is not producing the same output as Lotus (see https://github.com/ChainSafe/forest/actions/runs/11831922981/job/32968391970). I recommend running locally https://github.com/ChainSafe/forest/blob/main/scripts/tests/api_compare/setup.sh and ensuring the logic matches.

image

**Comment by virajbhartiya** [comment]
While running the API compare command, I came across a few more discrepancies,

```bash 
RUST_LOG=debug ./target/debug/forest-tool api compare --lotus /ip4/127.0.0.1/tcp/1234/http --forest /ip4/127.0.0.1/tcp/2345/http ../snap.zst --filter EthGetBlockReceipts -n 100
```
The discrepancies I could find:
```bash
Forest response: [
  {
    "blockHash": "0xed55e11735035e3e3a515a652456e0ba695348a7a7355ed987dac1a7dfdfb1d5",
    "blockNumber": "0x20b8ce",
    "contractAddress": null,
    "cumulativeGasUsed": "0x0",
    "effectiveGasPrice": "0x1eb21",
    "from": "0xff00000000000000000000000000000000000f03",
    "gasUsed": "0x579c29",
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "root": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "to": "0xff0000000000000000000000ffffffffffffffff",
    "transactionHash": "0xda75f460c1ec86a3dd126cce02be528c8c516516c94a78d51b17aa49f41d6cfa",
    "transactionIndex": "0x0",
    "type": "0x2"
  },
  {
    "blockHash": "0xed55e11735035e3e3a515a652456e0ba695348a7a7355ed987dac1a7dfdfb1d5",
    "blockNumber": "0x20b8ce",
    "contractAddress": null,
    "cumulativeGasUsed": "0x0",
    "effectiveGasPrice": "0x1e681",
    "from": "0xff00000000000000000000000000000000000f03",
    "gasUsed": "0x579c29",
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "root": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "to": "0xff0000000000000000000000ffffffffffffffff",
    "transactionHash": "0xe438c6d5f80857a2a2138fe7ec43a0695525b8670a498a201c5b64f71daadc5e",
    "transactionIndex": "0x1",
    "type": "0x2"
  }
]
Lotus response: [
  {
    "blockHash": "0xed55e11735035e3e3a515a652456e0ba695348a7a7355ed987dac1a7dfdfb1d5",
    "blockNumber": "0x20b8ce",
    "contractAddress": null,
    "cumulativeGasUsed": "0x0",
    "effectiveGasPrice": "0x1eb21",
    "from": "0xff00000000000000000000000000000000000f03",
    "gasUsed": "0x579c29",
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "root": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "to": "0xff00000000000000000000000000000000023fcc",
    "transactionHash": "0xda75f460c1ec86a3dd126cce02be528c8c516516c94a78d51b17aa49f41d6cfa",
    "transactionIndex": "0x0",
    "type": "0x2"
  },
  {
    "blockHash": "0xed55e11735035e3e3a515a652456e0ba695348a7a7355ed987dac1a7dfdfb1d5",
    "blockNumber": "0x20b8ce",
    "contractAddress": null,
    "cumulativeGasUsed": "0x0",
    "effectiveGasPrice": "0x1e681",
    "from": "0xff00000000000000000000000000000000000f03",
    "gasUsed": "0x579c29",
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "root": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "to": "0xff00000000000000000000000000000000023fcd",
    "transactionHash": "0xe438c6d5f80857a2a2138fe7ec43a0695525b8670a498a201c5b64f71daadc5e",
    "transactionIndex": "0x1",
    "type": "0x2"
  }
]
Diff:  [

   {

     "blockHash": "0xed55e11735035e3e3a515a652456e0ba695348a7a7355ed987dac1a7dfdfb1d5",

     "blockNumber": "0x20b8ce",

     "contractAddress": null,

     "cumulativeGasUsed": "0x0",

     "effectiveGasPrice": "0x1eb21",

     "from": "0xff00000000000000000000000000000000000f03",

     "gasUsed": "0x579c29",

     "logs": [],

     "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",

     "root": "0x0000000000000000000000000000000000000000000000000000000000000000",

     "status": "0x1",

-    "to": "0xff0000000000000000000000ffffffffffffffff",

+    "to": "0xff00000000000000000000000000000000023fcc",

     "transactionHash": "0xda75f460c1ec86a3dd126cce02be528c8c516516c94a78d51b17aa49f41d6cfa",

     "transactionIndex": "0x0",

     "type": "0x2"

   },

   {

     "blockHash": "0xed55e11735035e3e3a515a652456e0ba695348a7a7355ed987dac1a7dfdfb1d5",

     "blockNumber": "0x20b8ce",

     "contractAddress": null,

     "cumulativeGasUsed": "0x0",

     "effectiveGasPrice": "0x1e681",

     "from": "0xff00000000000000000000000000000000000f03",

     "gasUsed": "0x579c29",

     "logs": [],

     "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",

     "root": "0x0000000000000000000000000000000000000000000000000000000000000000",

     "status": "0x1",

-    "to": "0xff0000000000000000000000ffffffffffffffff",

+    "to": "0xff00000000000000000000000000000000023fcd",

     "transactionHash": "0xe438c6d5f80857a2a2138fe7ec43a0695525b8670a498a201c5b64f71daadc5e",

     "transactionIndex": "0x1",

     "type": "0x2"

   }

 ]
```
The block hash and number I could Fix but 

The Final result:
```bash
| RPC Method                        | Forest            | Lotus                                                                                                                                                              |
|-----------------------------------|-------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Filecoin.EthGetBlockReceipts      | CustomCheckFailed | Valid                                                                                                                                                              |
| Filecoin.EthGetBlockReceipts (19) | Valid             | Rejected("failed to create Eth receipt: failed to get eth logs for block and transaction: failed to get events for filter from chain indexer: not found in index") |
| Filecoin.EthGetBlockReceipts (80) | Valid             | Valid                                                   
```

**Comment by elmattic** [comment]
What version of Lotus are you using?

**Comment by virajbhartiya** [comment]
I am currently running `1.31.1-dev+calibnet`

**Comment by elmattic** [comment]
I will open a PR that fixes the `to` address mismatch.

**Comment by virajbhartiya** [comment]
Yes I did backfill the indexer db

**Comment by elmattic** [comment]
> I am currently running `1.31.1-dev+calibnet`

Note that on the CI we're still running `v1.30.0-rc1` so your mileage may vary.

**Comment by elmattic** [comment]
There’s a lint error in the CI ❌ that needs fixing. Can you enable 'format on save' in your editor? Thanks!

**Comment by virajbhartiya** [comment]
The CI is ✅ now!

**Review by elmattic — APPROVED** [review]

**Pull request merged**